README
NAME
    Devel::Symdump - dump symbol names or the symbol table

SYNOPSIS
        # Constructor
        require Devel::Symdump;
        @packs = qw(some_package another_package);
        $obj = Devel::Symdump->new(@packs);        # no recursion
        $obj = Devel::Symdump->rnew(@packs);       # with recursion

        # Methods
        @array = $obj->packages;
        @array = $obj->scalars;
        @array = $obj->arrays;
        @array = $obj->hashes;
        @array = $obj->functions;
        @array = $obj->filehandles;  # deprecated, use ios instead
        @array = $obj->dirhandles;   # deprecated, use ios instead
        @array = $obj->ios;
        @array = $obj->unknowns;     # only perl version < 5.003 had some

        $string = $obj->as_string;
        $string = $obj->as_HTML;
        $string = $obj1->diff($obj2);

        $string = Devel::Symdump->isa_tree;    # or $obj->isa_tree
        $string = Devel::Symdump->inh_tree;    # or $obj->inh_tree

        # Methods with autogenerated objects
        # all of those call new(@packs) internally
        @array = Devel::Symdump->packages(@packs);
        @array = Devel::Symdump->scalars(@packs);
        @array = Devel::Symdump->arrays(@packs);
        @array = Devel::Symdump->hashes(@packs);
        @array = Devel::Symdump->functions(@packs);
        @array = Devel::Symdump->ios(@packs);
        @array = Devel::Symdump->unknowns(@packs);

DESCRIPTION
    This little package serves to access the symbol table of perl.

    "Devel::Symdump->rnew(@packages)"
        returns a symbol table object for all subtrees below @packages.
        Nested Modules are analyzed recursively. If no package is given as
        argument, it defaults to "main". That means to get the whole symbol
        table, just do a "rnew" without arguments.

        The global variable $Devel::Symdump::MAX_RECURSION limits the
        recursion to prevent contention. The default value is set to 97,
        just low enough to survive the test suite without a warning about
        deep recursion.

    "Devel::Symdump->new(@packages)"
        does not go into recursion and only analyzes the packages that are
        given as arguments.

    packages, scalars, arrays, hashes, functions, ios
        The methods packages(), scalars(), arrays(), hashes(), functions(),
        ios(), and (for older perls) unknowns() each return an array of
        fully qualified symbols of the specified type in all packages that
        are held within a Devel::Symdump object, but without the leading
        "$", "@" or "%". In a scalar context, they will return the number of
        such symbols. Unknown symbols are usually either formats or
        variables that haven't yet got a defined value.

        Note that scalar symbol table entries are a special case. If a
        symbol table entry exists at all, presence of a scalar is currently
        unknowable, due to a feature of Perl described in perlref "Making
        References" point 7. For example, this package will mark a scalar
        value $foo as present if any of @foo, %foo, &foo etc. have been
        declared or used.

    as_string
    as_HTML
        As_string() and as_HTML() return a simple string/HTML
        representations of the object.

    diff
        Diff() prints the difference between two Devel::Symdump objects in
        human readable form. The format is similar to the one used by the
        as_string method.

    isa_tree
    inh_tree
        Isa_tree() and inh_tree() both return a simple string representation
        of the current inheritance tree. The difference between the two
        methods is the direction from which the tree is viewed: top-down or
        bottom-up. As I'm sure, many users will have different expectation
        about what is top and what is bottom, I'll provide an example what
        happens when the Socket module is loaded:

    % print Devel::Symdump->inh_tree
            AutoLoader
                    DynaLoader
                            Socket
            DynaLoader
                    Socket
            Exporter
                    Carp
                    Config
                    Socket

        The inh_tree method shows on the left hand side a package name and
        indented to the right the packages that use the former.

    % print Devel::Symdump->isa_tree
            Carp
                    Exporter
            Config
                    Exporter
            DynaLoader
                    AutoLoader
            Socket
                    Exporter
                    DynaLoader
                            AutoLoader

        The isa_tree method displays from left to right ISA relationships,
        so Socket IS A DynaLoader and DynaLoader IS A AutoLoader. (Actually,
        they were at the time this manpage was written)

    You may call both methods, isa_tree() and inh_tree(), with an object. If
    you do that, the object will store the output and retrieve it when you
    call the same method again later. The typical usage would be to use them
    as class methods directly though.

SUBCLASSING
    The design of this package is intentionally primitive and allows it to
    be subclassed easily. An example of a (maybe) useful subclass is
    Devel::Symdump::Export, a package which exports all methods of the
    Devel::Symdump package and turns them into functions.

SEE ALSO
    Package::Stash

AUTHORS
    Andreas Koenig <andk@cpan.org> and Tom Christiansen <tchrist@perl.com>.
    Based on the old dumpvar.pl by Larry Wall.

COPYRIGHT, LICENSE
    This module is

    Copyright (c) 1995, 1997, 2000, 2002, 2005, 2006 Andreas Koenig
    "<andk@cpan.org>".

    All rights reserved.

    This library is free software; you may use, redistribute and/or modify
    it under the same terms as Perl itself.



Changes
2014-12-16  k  <andk@cpan.org>

	* release 2.14

	* no change to 2.14-TRIAL

2014-12-08  k  <andk@cpan.org>

	* release 2.14-TRIAL

	* support for fileno() on directory handles (perl commit v5.21.6-337-g67f2cc7)

2014-07-26  k  <andk@cpan.org>

	* release 2.13-TRIAL

	* support for telldir (Brian Fraser)

2014-06-20  k  <andk@cpan.org>

	* release 2.12

	* test rewrite: v5.21.0-424-ge35475d stopped supporting
	defined(@$ref) which was used in t/symdump.t (thanks to Aaron
	Crane for spotting)

2013-10-30  k  <andk@cpan.org>

	* release 2.11

	* v5.19.5-71-gd456e3f stopped producing the %@ hash at startup

2013-03-27  k  <k@k83.linux.bogus>

	* release 2.10

	* no change to 2.10-TRIAL

2013-03-24  k  <k@k83.linux.bogus>

	* release 2.10-TRIAL

	* release 2.09-TRIAL

	* address RT#84139: fix test to work with upcoming perl 5.18
	(Thanks to ZEFRAM)

	* fixed the autogen.t test again, this time for perl 5.8.9

2012-05-20  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* release 2.08_53

	* apply doc patch by Nick Stokoe from ticket #77102

	* declare dependency on Compress::Zlib

2009-03-01  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* release 2.08_51

	* added a test by Jason M. Mills to chase down the bug he reports
	in https://rt.cpan.org/Ticket/Display.html?id=43675

2007-10-11  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* release 2.08

	* skip the recurse test on 5.005

2007-01-05  Andreas J. Koenig  <andreas.koenig.7os6VVqR@franz.ak.mind.de>

	* release 2.07

	* rewrite symdump.t using Test::More

	* adjust test suite to accept main::- also introduced for named
	captures

2006-10-08  Andreas J. Koenig  <andreas.koenig.gmwojprw@franz.ak.mind.de>

	* release 2.0604

	* adjust test suite to accept main::+ introduced by named captures

2006-09-20  Andreas J. Koenig  <andreas.koenig.gmwojprw@franz.ak.mind.de>

	* relase 2.0603

	* add LICENSE field to Makefile.PL to also have it in the META.yml

2006-07-19  Andreas J. Koenig  <andreas.koenig.gmwojprw@franz.ak.mind.de>

	* release 2.0602

	* adjust test suite to accept new variable in bleadperl, the hash
	$main::^H

2006-05-03  Andreas J. Koenig  <andreas.koenig.gmwojprw@franz.ak.mind.de>

	* release 2.0601

	* add copyright and license

2006-01-18  Andreas J. Koenig  <andreas.koenig.gmwojprw@franz.ak.mind.de>

	* release 2.06

	* New warnings in bleadperl now suppressed; minor pod issues fixed

2006-01-02  Andreas J. Koenig  <andreas.koenig.gmwojprw@franz.ak.mind.de>

	* release 2.05

	* fix the testcase for recursion so that it compiles and works
	also after patch 26370 to perl after which stashes are not
	autovivified anymore.

2005-12-25  Andreas J. Koenig  <andreas.koenig.gmwojprw@franz.ak.mind.de>

	* release 2.04

	* Export.pm now strict clean

	* Makefile.PL up to date

	* added ChangeLog.svn

	* added tests for recursion, pod, podcover

	* Fixed rt.cpan.org #8766--recursion

2002-03-01  Andreas J. Koenig  <andreas.koenig@anima.de>

	* lib/Devel/Symdump.pm: perl 5.6.1 introduced a package name of
	"<none>" to work around a bug if somebody uses the deprecated
	C<package> without an argument. I believe we need to ignore that
	symbol completely so that at least we follow the lead of the B::
	extensions. Thanks to Sreeji K Das /sreeji_k at yahoo.com/ for the
	report.

2000-10-31  Andreas J. Koenig  <andreas.koenig@anima.de>

	* Typo fix: hashs --> hashes. Thanks to Sebastien Blondeel
	<Sebastien.Blondeel@idealx.com> for the report.

2000-06-14  Andreas J. Koenig  <andreas.koenig@anima.de>

	* Fixed my email address in and made a few tiny editorial changes
	to the manpage.

	* Replaced Changes file with this ChangeLog file, appended the
	full Changes file below.

	* Fixed the test 6 in t/symdump.t. This test was broken by
	perl-5.6.0 but not Devel::Symdump itself.

1997-05-16  Andreas Koenig

	* Release 2.00

	* Fixed typos in the manpage, added a test for tree, no functional
	change, released 2.00.

1997-03-31  Andreas Koenig

	* 1.99_01

	* 1.99_01 is the designated 2.00.

	* Between 1.20 and 1.23 the method as_HTML was introduced and a
	few code cleanups happened.

	* 2.00 switches implementation to use *ENTRY{XXX} internally. This
	means that we can determine scalarness even for undefined scalars.
	We don't expect unknowns anymore.

	* 2.00 introduces the new ios() method which should replace the
	older filehandles() and dirhandles() methods. For backwards
	compatibility the old methods continue to work as they used to.

	* 2.00 comes with isa_tree and inh_tree utility methods for
	analysing the inheritance tree. Devel::Symdump objects may be used
	to create snapshots, but their typical use would be as class
	methods.

1995-08-16  Andreas Koenig

	* 1.20

	* test 7 of t/symdump.t was too capricious. In fact the test was
	based on wrong assumptions about loaded packages in the perl
	binary. Static perls and dynamic perls have different symbol
	tables when they run thetests. So test 7 is gone.

	* test 4 of t/symdump.t relied on $@ being set like in perl5.001m.
	This might not be a correct assumption. So test 4 is replaced with
	a dummy 'print ok' until the $@ problem is sorted out

1995-07-03  Andreas Koenig

	* 1.19

	* Added an as_string method.

	* Rewrote the test scripts so they output standard test strings
	"ok nnn". This is dangerous for new perl releases, but will help
	me do get bug reports early.

	* Renamed the exporting example package to Devel::Symdump::Export.

1995-05-29  Andreas Koenig

	* 1.16

	* Changed '${pack}::' and relatives back to "$pack\:\:" to make
	the package "-w" safe. Deleted the debug statement in _doit()
	after Gurusamy Sarathy fixed the bug in perl5.001, but left a
	comment there.

	* Changed the AUTHORS section to plain "Andreas & Tom".

	* Added this Changes file :)

1995-05-28  Andreas Koenig

	* 1.14

	* After a considerable amount of mail exchange between Tom and me,
	we now have a unknowns() method for all the rest in the symbol

Makefile.PL
#!/usr/bin/perl -w -*- mode: cperl -*-
use strict;
use vars qw( $VERSION );
use ExtUtils::MakeMaker qw(:DEFAULT);

eval { require File::Spec; };
my $HAVE_FILE_SPEC = !$@;

my $version_diff = 0; # we'll have to die if this becomes true
my $version_from;
my $is_trial = 1;
my $version;
{
  local $^W;
  $ExtUtils::MakeMaker::VERSION = eval $ExtUtils::MakeMaker::VERSION;
}
if ($HAVE_FILE_SPEC) {
  $version_from = File::Spec->catfile(qw(lib Devel Symdump.pm));
} else {
  $version_from = q(lib/Devel/Symdump.pm);
}
unshift @INC, "lib";
require $version_from;
$version = $Devel::Symdump::VERSION;
if ($is_trial && $version !~ /_/) {
    $version .= "-TRIAL";
}
{
  my $version_set_manually = 1; # not by SVN
  if ($ARGV[0] && $ARGV[0] eq "--setversion") {
    die "Your perl is a bit dated[$]].\nDo not make a release with it\n" if $] < 5.008;
    die "Your MakeMaker is a bit dated[$ExtUtils::MakeMaker::VERSION].\nDo not make a release with it\n"
        if $ExtUtils::MakeMaker::VERSION < 6.4502;
    die "Your MakeMaker doesn't do the sign woodoo" unless
        MM->can("signature_target");
    shift @ARGV;
    my $st;
    local $ENV{LANG} = "C";
    my $dirty = `git status --porcelain --untracked-files=no`;
    die "Not everything checked in or out?\n====\n$dirty====\n" if $dirty;

    if ($version_set_manually) {
      # we must control that the VERSION in this .pm is the same as in the Makefile
      open my $fh, "make the-release-name|" or die;
      my $have_version;
      while (<$fh>) {
        next unless /^version\s+([\d\._]+(?:-TRIAL)?)/;
        $have_version = $1;
      }
      die "could not determine current version from Makefile" unless $have_version;
      eval q{
      no warnings "numeric";
      my $dsv = $Devel::Symdump::VERSION;
      if ($dsv != $have_version) {
        warn "Not equal: D:S:VERSION[$dsv] Makefile version[$have_version]";
        $version_diff = 1;
      }
};
      die $@ if $@;
    }
    exit unless $version_diff;
  }
}

my $prereq_pm = {
                 'Compress::Zlib' => 0, # only for t/glob_to_local_typeglob.t
                 'Test::More' => 0,
                };
my @sign = (MM->can("signature_target") ? (SIGN => 1) : ());
WriteMakefile(
	      NAME => "Devel::Symdump",
	      DISTNAME => "Devel-Symdump",
	      VERSION => $version,
              PREREQ_PM    => $prereq_pm,
              ($ExtUtils::MakeMaker::VERSION >= 6.3002 ?
               (LICENSE      => "perl") : (),
              ),
              ($ExtUtils::MakeMaker::VERSION >= 6.48 ?
               (MIN_PERL_VERSION => '5.004') : (),
              ),
              clean => {
                        FILES => '*/*/*~',
                       },
              @sign,
              ($] >= 5.005 ?
               (
                ABSTRACT_FROM  => 'lib/Devel/Symdump.pm', # retrieve abstract from module
                AUTHOR         => 'Andreas Koenig <andreas.koenig.gmwojprw@franz.ak.mind.de>') : (),
              ),
              dist => {
                       DIST_DEFAULT => join(" ", # note: order matters!
                                            "verify-no-subdir",
                                            "verify-changes-date",
                                            "verify-changes-version",
                                            "Makefile",
                                            "setversion",
                                            "README",
                                            "all",
                                            "tardist",
                                           ),
                       COMPRESS => 'gzip -9',
                      },
              # I took it from RT-CPAN ticket 30098:
              ($ExtUtils::MakeMaker::VERSION >= 6.4502 ?
               (META_ADD => {
                             resources => {
                                 repository => "git://github.com/andk/devel-symdump.git",
                             },
                             keywords => ['symbol table inspection'],
                            }) : ()),
             );

if ($version_diff){
  die "
==> I had to update some \$VERSIONs <==
==> Your Makefile has been rebuilt. <==
==> Please rerun the make command.  <==
";
}

package MY;
sub distsignature {
    my($self) = shift;
    my $ret = $self->SUPER::distsignature_target(@_);
    $ret =~ s|cpansign|\`dirname \$(PERL)\`/cpansign|g;
    return $ret;
}
sub macro {
    q{
LC_ALL_noexport=en_GB.utf8

YAML_MODULE=YAML::Syck
}
}

sub postamble {
  q{
# the subdirs on MY OWN BOX are allowed here (only used for make dist!)
OKDIRS=benchmark|bin|blib|lib|scripts|t

verify-no-subdir:
	@$(PERL) -e 'my$$s=join",",grep{!/^($(OKDIRS))\z/x&&-d($$_)}glob"*";' \
		-e 'die"unexpected dir:$$s"if$$s'

verify-changes-date:
	@$(PERL) -ne 'BEGIN{my@t=(localtime)[5,4,3];$$t[0]+=1900;$$t[1]++;$$t=sprintf"%04d-%02d-%02d",@t}' \
		-e '$$ok++,exit if /^$$t\s/; END{die "Alert: did not find <$$t> in Changes file" unless $$ok}' Changes

verify-changes-version:
	@$(PERL) -ne '$$ok++,exit if /\b$(VERSION)\b/; END{die "Alert: did not find <$(VERSION)> in Changes file" unless $$ok}' Changes

setversion:
	$(PERL) Makefile.PL --setversion

README: lib/Devel/Symdump.pm Makefile
	-test -r $@ && chmod +w $@
	-$(PERL) -MPod::Text -e 'Pod::Text->new->parse_from_file(\*ARGV)' lib/Devel/Symdump.pm > $@

the-release-name :
	$(NOECHO) $(ECHO) 'version ' $(VERSION)
	$(NOECHO) $(ECHO) 'release-name ' $(DISTVNAME).tar$(SUFFIX)

release :: disttest
	git tag -m 'This is $(VERSION)' "$(VERSION)"
	ls -l $(DISTVNAME).tar$(SUFFIX)
	rm -rf $(DISTVNAME)
	$(NOECHO) $(ECHO) '#### Suggested next steps:'
	$(NOECHO) $(ECHO) '  git push --tags origin master'

sign:
	cpansign -s

howto-release:
	@$(ECHO) manually set version in Symdump.pm, edit ChangeLog
	@$(ECHO) make ci dist \&\& make release
}
}

sub dist_ci {
  return qq{ci :
	svn ci
};
}

sub dist_test {
  return q{
# if we depend on $(DISTVNAME).tar$(SUFFIX), then the rest of the
# Makefile breaks our intent to NOT remake dist
disttest :
	rm -rf $(DISTVNAME)
	tar xvzf $(DISTVNAME).tar$(SUFFIX)
	cd $(DISTVNAME) && $(ABSPERLRUN) Makefile.PL
	cd $(DISTVNAME) && $(MAKE) $(PASTHRU)
	cd $(DISTVNAME) && $(MAKE) test $(PASTHRU)

distdir ::
	touch $(DISTVNAME)/SIGNATURE && $(CP) $(DISTVNAME)/SIGNATURE ./SIGNATURE
	$(CP) $(DISTVNAME)/META.yml  ./META.yml
	$(CP) $(DISTVNAME)/META.json  ./META.json
	$(CP) $(DISTVNAME)/MANIFEST  ./MANIFEST

}
}

sub distdir {
  my $self = shift;
  my $out = $self->SUPER::distdir;
  $out =~ s/distdir :/distdir ::/g;
  return $out;
}

# dist_dir was the name in very old MakeMaker as of 5.005_04
sub dist_dir {
  my $self = shift;
  my $out = $self->SUPER::dist_dir;
  $out =~ s/distdir :/distdir ::/g;
  return $out;
}


lib/Devel/Symdump.pm
package Devel::Symdump;

use 5.003;
use Carp ();
use strict;
use vars qw($Defaults $VERSION *ENTRY $MAX_RECURSION);

$VERSION = '2.14';
$MAX_RECURSION = 97;

$Defaults = {
	     'RECURS'   => 0,
	     'AUTOLOAD' => {
			    'packages'	=> 1,
			    'scalars'	=> 1,
			    'arrays'	=> 1,
			    'hashes'	=> 1,
			    'functions'	=> 1,
			    'ios'	=> 1,
			    'unknowns'	=> 1,
			   },
             'SEEN' => {},
	    };

sub rnew {
    my($class,@packages) = @_;
    no strict "refs";
    my $self = bless {%${"$class\::Defaults"}}, $class;
    $self->{RECURS}++;
    $self->_doit(@packages);
}

sub new {
    my($class,@packages) = @_;
    no strict "refs";
    my $self = bless {%${"$class\::Defaults"}}, $class;
    $self->_doit(@packages);
}

sub _doit {
    my($self,@packages) = @_;
    @packages = ("main") unless @packages;
    $self->{RESULT} = $self->_symdump(@packages);
    return $self;
}

sub _symdump {
    my($self,@packages) = @_ ;
    my($key,$val,$num,$pack,@todo,$tmp);
    my $result = {};
    foreach $pack (@packages){
	no strict;
	while (($key,$val) = each(%{*{"$pack\::"}})) {
	    my $gotone = 0;
	    local(*ENTRY) = $val;
	    #### SCALAR ####
	    if (defined $val && defined *ENTRY{SCALAR}) {
		$result->{$pack}{SCALARS}{$key}++;
		$gotone++;
	    }
	    #### ARRAY ####
	    if (defined $val && defined *ENTRY{ARRAY}) {
		$result->{$pack}{ARRAYS}{$key}++;
		$gotone++;
	    }
	    #### HASH ####
	    if (defined $val && defined *ENTRY{HASH} && $key !~ /::/) {
		$result->{$pack}{HASHES}{$key}++;
		$gotone++;
	    }
	    #### PACKAGE ####
	    if (defined $val && defined *ENTRY{HASH} && $key =~ /::$/ &&
                $key ne "main::" && $key ne "<none>::") {
                my($p) = $pack ne "main" ? "$pack\::" : "";
                ($p .= $key) =~ s/::$//;
                $result->{$pack}{PACKAGES}{$p}++;
                $gotone++;
                if (++$self->{SEEN}{*$val} > $Devel::Symdump::MAX_RECURSION){
                    next;
                }
		push @todo, $p;
	    }
	    #### FUNCTION ####
	    if (defined $val && defined *ENTRY{CODE}) {
		$result->{$pack}{FUNCTIONS}{$key}++;
		$gotone++;
	    }

	    #### IO #### had to change after 5.003_10
	    if ($] > 5.003_10){
		if (defined $val && defined *ENTRY{IO}){ # fileno and telldir...
		    $result->{$pack}{IOS}{$key}++;
		    $gotone++;
		}
	    } else {
		#### FILEHANDLE ####
		if (defined fileno(ENTRY)){
		    $result->{$pack}{IOS}{$key}++;
		    $gotone++;
		} elsif (defined telldir(ENTRY)){
		    #### DIRHANDLE ####
		    $result->{$pack}{IOS}{$key}++;
		    $gotone++;
		}
	    }

	    #### SOMETHING ELSE ####
	    unless ($gotone) {
		$result->{$pack}{UNKNOWNS}{$key}++;
	    }
	}
    }

    return (@todo && $self->{RECURS})
		? { %$result, %{$self->_symdump(@todo)} }
		: $result;
}

sub _partdump {
    my($self,$part)=@_;
    my ($pack, @result);
    my $prepend = "";
    foreach $pack (keys %{$self->{RESULT}}){
	$prepend = "$pack\::" unless $part eq 'PACKAGES';
	push @result, map {"$prepend$_"} keys %{$self->{RESULT}{$pack}{$part} || {}};
    }
    return @result;
}

# this is needed so we don't try to AUTOLOAD the DESTROY method
sub DESTROY {}

sub as_string {
    my $self = shift;
    my($type,@m);
    for $type (sort keys %{$self->{'AUTOLOAD'}}) {
	push @m, $type;
	push @m, "\t" . join "\n\t", map {
	    s/([\000-\037\177])/ '^' . pack('c', ord($1) ^ 64) /eg;
	    $_;
	} sort $self->_partdump(uc $type);
    }
    return join "\n", @m;
}

sub as_HTML {
    my $self = shift;
    my($type,@m);
    push @m, "<TABLE>";
    for $type (sort keys %{$self->{'AUTOLOAD'}}) {
	push @m, "<TR><TD valign=top><B>$type</B></TD>";
	push @m, "<TD>" . join ", ", map {
	    s/([\000-\037\177])/ '^' .
		pack('c', ord($1) ^ 64)
		    /eg; $_;
	} sort $self->_partdump(uc $type);
	push @m, "</TD></TR>";
    }
    push @m, "</TABLE>";
    return join "\n", @m;
}

sub diff {
    my($self,$second) = @_;
    my($type,@m);
    for $type (sort keys %{$self->{'AUTOLOAD'}}) {
	my(%first,%second,%all,$symbol);
	foreach $symbol ($self->_partdump(uc $type)){
	    $first{$symbol}++;
	    $all{$symbol}++;
	}
	foreach $symbol ($second->_partdump(uc $type)){
	    $second{$symbol}++;
	    $all{$symbol}++;
	}
	my(@typediff);
	foreach $symbol (sort keys %all){
	    next if $first{$symbol} && $second{$symbol};
	    push @typediff, "- $symbol" unless $second{$symbol};
	    push @typediff, "+ $symbol" unless $first{$symbol};
	}
	foreach (@typediff) {
	    s/([\000-\037\177])/ '^' . pack('c', ord($1) ^ 64) /eg;
	}
	push @m, $type, @typediff if @typediff;
    }
    return join "\n", @m;
}

sub inh_tree {
    my($self) = @_;
    return $self->{INHTREE} if ref $self && defined $self->{INHTREE};
    my($inherited_by) = {};
    my($m)="";
    my(@isa) = grep /\bISA$/, Devel::Symdump->rnew->arrays;
    my $isa;
    foreach $isa (sort @isa) {
	$isa =~ s/::ISA$//;
	my($isaisa);
	no strict 'refs';
	foreach $isaisa (@{"$isa\::ISA"}){
	    $inherited_by->{$isaisa}{$isa}++;
	}
    }
    my $item;
    foreach $item (sort keys %$inherited_by) {
	$m .= "$item\n";
	$m .= _inh_tree($item,$inherited_by);
    }
    $self->{INHTREE} = $m if ref $self;
    $m;
}

sub _inh_tree {
    my($package,$href,$depth) = @_;
    return unless defined $href;
    $depth ||= 0;
    $depth++;
    if ($depth > 100){
	warn "Deep recursion in ISA\n";
	return;
    }
    my($m) = "";
    # print "DEBUG: package[$package]depth[$depth]\n";
    my $i;
    foreach $i (sort keys %{$href->{$package}}) {
	$m .= qq{\t} x $depth;
	$m .= qq{$i\n};
	$m .= _inh_tree($i,$href,$depth);
    }
    $m;
}

sub isa_tree{
    my($self) = @_;
    return $self->{ISATREE} if ref $self && defined $self->{ISATREE};
    my(@isa) = grep /\bISA$/, Devel::Symdump->rnew->arrays;
    my($m) = "";
    my($isa);
    foreach $isa (sort @isa) {
	$isa =~ s/::ISA$//;
	$m .= qq{$isa\n};
	$m .= _isa_tree($isa)
    }
    $self->{ISATREE} = $m if ref $self;
    $m;
}

sub _isa_tree{
    my($package,$depth) = @_;
    $depth ||= 0;
    $depth++;
    if ($depth > 100){
	warn "Deep recursion in ISA\n";
	return;
    }
    my($m) = "";
    # print "DEBUG: package[$package]depth[$depth]\n";
    my $isaisa;
    no strict 'refs';
    foreach $isaisa (@{"$package\::ISA"}) {
	$m .= qq{\t} x $depth;
	$m .= qq{$isaisa\n};
	$m .= _isa_tree($isaisa,$depth);
    }
    $m;
}

AUTOLOAD {
    my($self,@packages) = @_;
    unless (ref $self) {
	$self = $self->new(@packages);
    }
    no strict "vars";
    (my $auto = $AUTOLOAD) =~ s/.*:://;

    $auto =~ s/(file|dir)handles/ios/;
    my $compat = $1;

    unless ($self->{'AUTOLOAD'}{$auto}) {
	Carp::croak("invalid Devel::Symdump method: $auto()");
    }

    my @syms = $self->_partdump(uc $auto);
    if (defined $compat) {
	no strict 'refs';
        local $^W; # bleadperl@26631 introduced an io warning here
	if ($compat eq "file") {
	    @syms = grep { defined(fileno($_)) } @syms;
	} else {
	    @syms = grep { _is_dirhandle($_) } @syms;
	}
    }
    return @syms; # make sure now it gets context right
}

use Config ();
use constant HAVE_TELLDIR => $Config::Config{d_telldir};
sub _is_dirhandle {
    my ($glob) = @_;
    if ( HAVE_TELLDIR ) {
        return defined(telldir($glob));
    }
    else {
        if ( !ref $glob ) {
            no strict 'refs';
            $glob = \*{$glob};
        }
        require B;
        my $obj = B::svref_2object($glob);
        return if !$obj || !eval{ $obj->IO; $obj->IO->IoTYPE; 1 };
        my $mode = $obj->IO->IoTYPE;
        return $mode eq "\0" ? 1 : 0;
    }
}

1;

__END__

=head1 NAME

Devel::Symdump - dump symbol names or the symbol table

=head1 SYNOPSIS

    # Constructor
    require Devel::Symdump;
    @packs = qw(some_package another_package);
    $obj = Devel::Symdump->new(@packs);        # no recursion
    $obj = Devel::Symdump->rnew(@packs);       # with recursion

    # Methods
    @array = $obj->packages;
    @array = $obj->scalars;
    @array = $obj->arrays;
    @array = $obj->hashes;
    @array = $obj->functions;
    @array = $obj->filehandles;  # deprecated, use ios instead
    @array = $obj->dirhandles;   # deprecated, use ios instead
    @array = $obj->ios;
    @array = $obj->unknowns;     # only perl version < 5.003 had some

    $string = $obj->as_string;
    $string = $obj->as_HTML;
    $string = $obj1->diff($obj2);

    $string = Devel::Symdump->isa_tree;    # or $obj->isa_tree
    $string = Devel::Symdump->inh_tree;    # or $obj->inh_tree

    # Methods with autogenerated objects
    # all of those call new(@packs) internally
    @array = Devel::Symdump->packages(@packs);
    @array = Devel::Symdump->scalars(@packs);
    @array = Devel::Symdump->arrays(@packs);
    @array = Devel::Symdump->hashes(@packs);
    @array = Devel::Symdump->functions(@packs);
    @array = Devel::Symdump->ios(@packs);
    @array = Devel::Symdump->unknowns(@packs);

=head1 DESCRIPTION

This little package serves to access the symbol table of perl.

=over 4

=item C<Devel::Symdump-E<gt>rnew(@packages)>

returns a symbol table object for all subtrees below @packages.
Nested Modules are analyzed recursively. If no package is given as
argument, it defaults to C<main>. That means to get the whole symbol
table, just do a C<rnew> without arguments.

The global variable $Devel::Symdump::MAX_RECURSION limits the
recursion to prevent contention. The default value is set to 97, just
low enough to survive the test suite without a warning about deep
recursion.

=item C<Devel::Symdump-E<gt>new(@packages)>

does not go into recursion and only analyzes the packages that are
given as arguments.

=item packages, scalars, arrays, hashes, functions, ios

The methods packages(), scalars(), arrays(), hashes(), functions(),
ios(), and (for older perls) unknowns() each return an array of fully
qualified symbols of the specified type in all packages that are held
within a Devel::Symdump object, but without the leading C<$>, C<@> or
C<%>. In a scalar context, they will return the number of such
symbols. Unknown symbols are usually either formats or variables that
haven't yet got a defined value.

Note that scalar symbol table entries are a special case.  If a symbol
table entry exists at all, presence of a scalar is currently
unknowable, due to a feature of Perl described in perlref "Making
References" point 7.  For example, this package will mark a scalar
value C<$foo> as present if any of C<@foo>, C<%foo>, C<&foo> etc. have
been declared or used.

=item as_string

=item as_HTML

As_string() and as_HTML() return a simple string/HTML representations
of the object.

=item diff

Diff() prints the difference between two Devel::Symdump objects in
human readable form. The format is similar to the one used by the
as_string method.

=item isa_tree

=item inh_tree

Isa_tree() and inh_tree() both return a simple string representation
of the current inheritance tree. The difference between the two
methods is the direction from which the tree is viewed: top-down or
bottom-up. As I'm sure, many users will have different expectation
about what is top and what is bottom, I'll provide an example what
happens when the Socket module is loaded:

=item % print Devel::Symdump-E<gt>inh_tree

    AutoLoader
            DynaLoader
                    Socket
    DynaLoader
            Socket
    Exporter
            Carp
            Config
            Socket

The inh_tree method shows on the left hand side a package name and
indented to the right the packages that use the former.

=item % print Devel::Symdump-E<gt>isa_tree

    Carp
            Exporter
    Config
            Exporter
    DynaLoader
            AutoLoader
    Socket
            Exporter
            DynaLoader
                    AutoLoader

The isa_tree method displays from left to right ISA relationships, so
Socket IS A DynaLoader and DynaLoader IS A AutoLoader. (Actually, they
were at the time this manpage was written)

=back

You may call both methods, isa_tree() and inh_tree(), with an
object. If you do that, the object will store the output and retrieve
it when you call the same method again later. The typical usage would
be to use them as class methods directly though.

=head1 SUBCLASSING

The design of this package is intentionally primitive and allows it to
be subclassed easily. An example of a (maybe) useful subclass is
Devel::Symdump::Export, a package which exports all methods of the
Devel::Symdump package and turns them into functions.

=head1 SEE ALSO

Routines for manipulating stashes: C<Package::Stash>; to work with
lexicals: C<PadWalker>.

=head1 AUTHORS

Andreas Koenig F<< <andk@cpan.org> >> and Tom Christiansen
F<< <tchrist@perl.com> >>. Based on the old F<dumpvar.pl> by Larry
Wall.

=head1 COPYRIGHT, LICENSE

This module is

Copyright (c) 1995, 1997, 2000, 2002, 2005, 2006 Andreas Koenig C<< <andk@cpan.org> >>.

All rights reserved.

This library is free software;
you may use, redistribute and/or modify it under the same
terms as Perl itself.

=cut


# Local Variables:
# mode: cperl
# cperl-indent-level: 4
# End:
